File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int n = board.size(); 
        vector<set<char>> rows(n); 
        vector<set<char>> cols(n); 
        vector<set<char>> boxes(n); 

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] == '.') continue; 

                char digit = board[i][j]; 

                int boxIndex = (i / 3) * 3 + (j / 3); 

                if(rows[i].count(digit) || cols[j].count(digit) || boxes[boxIndex].count(digit)){
                    return false; 
                }

                rows[i].insert(digit);
                cols[j].insert(digit); 
                boxes[boxIndex].insert(digit); 
            }
        }
        return true; 
    }
};

File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        Map<Integer,Set<Character>> rowMap = new HashMap<>(); 
        Map<Integer,Set<Character>> colMap = new HashMap<>(); 
        Map<Integer,Set<Character>> boxMap = new HashMap<>(); 

        for(int i = 0; i < 9; i++){
            for(int j = 0; j < 9; j++){
                if(board[i][j] == '.') continue; 
                char c = board[i][j]; 
                
                int boxIndex = (i / 3) * 3 + (j / 3); 

                if(rowMap.containsKey(i) && rowMap.get(i).contains(c)) return false; 
                if(colMap.containsKey(j) && colMap.get(j).contains(c)) return false; 
                if(boxMap.containsKey(boxIndex) && boxMap.get(boxIndex).contains(c)) return false; 

                rowMap.computeIfAbsent(i, k -> new HashSet<>()).add(c); 
                colMap.computeIfAbsent(j, k -> new HashSet<>()).add(c); 
                boxMap.computeIfAbsent(boxIndex, k -> new HashSet<>()).add(c); 
            }
        }

        return true; 
    }
}